Add WTF Python Command - Continuation of #605

brad90four opened the pull request:
## Relevant Issues
<!--
It is mandatory to link to an issue that has been approved by a Core Developer, indicated by an "approved" label.
Issues can be skipped with explicit core dev approval, but you have to link the discussion.
-->
Closes PR #605 
Closes Issue #602 

<!-- Link the issue by typing: "Closes #<number>" (Closes #0 to close issue 0 for example). -->


## Description
<!-- Describe what changes you made, and how you've implemented them. -->
Gets the fuzzy match of a search query, returns None if the ratio is below 90 else returns the fuzzy match. If the return is None, then we send our custom Error Message else it sends the link to the subsection of the WTF Python README.md.

## Reasoning
Approved by Staff

## Screenshots from original PR
![image](https://user-images.githubusercontent.com/42116429/132347900-6bdddd8f-7946-4dd7-bbb4-a21ab862b4c2.png)

## Did you:
<!-- These are required when contributing. -->
<!-- Replace [ ] with [x] to mark items as done. -->

- [X] Join the [**Python Discord Community**](https://discord.gg/python)?
- [X] Read all the comments in this template?
- [X] Ensure there is an issue open, or link relevant discord discussions?
- [X] Read and agree to the [contributing guidelines](https://pythondiscord.com/pages/contributing/contributing-guidelines/)?

Shom770 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,145 @@
+            embed = Embed(
+                title=random.choice(constants.ERROR_REPLIES),
+                description=ERROR_MESSAGE,
+                colour=constants.Colours.soft_red
```
Trailing comma,
whenever there are parentheses broken down, at the end there should always be a trailing comma as it makes the git blame more accurate, and adheres to PEP8
Like for this line, it would be:
```py
embed = Embed(
           title=random.choice(constants.ERROR_REPLIES),
           description=ERROR_MESSAGE,
           colour=constants.Colours.soft_red,
            )
```

Shom770 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,145 @@
+            embed = Embed(
+                title=f"WTF Python Search Result For {query}",
+                colour=constants.Colours.dark_green,
+                description=f"[Go to Repository Section]({self.headers[match]})"
```
See comment above regarding trailing comma

Shom770 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,145 @@
+
+
+def setup(bot: Bot) -> None:
+    """Load WTFPython Cog."""
```
Grammatical nitpick - ``"""Load the WTFPython Cog"""``

Shom770 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,145 @@
+                )
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fuzzy match of a query if its ratio is above "MINIMUM_CERTAINTY" else returns None."""
```
Could you give a brief explanation of what the ratio means in this context?
Possibly adding it to the docstring so if anyone ever needs to edit this in the future they know what it means?

Shom770 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,145 @@
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.get_WTF_python_readme.start()
```
There is a typo - this should be ``self.get_wtf_python_readme.start()``

Shom770 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,149 @@
+            action = Action.LOAD
+
+        verb = action.name.lower()
+        ext = "bot.exts.evergreen.wtf_python"
```
This won't work as the filestructure got updated, so this should be
```suggestion
        ext = "bot.exts.utilities.wtf_python"
```

Shom770 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,149 @@
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fuzzy match of a query if its ratio is above "MINIMUM_CERTAINTY" else returns None."""
+        match, certainty = process.extractOne(query, self.headers.keys())
```
This will return a ValueError, as rapidfuzz.process got updated for extractOne where it returns three values, the match itself, the certainty, and the index the string was found.
It doesn't seem like you are using the index so you could just replace it with a wildcard.
This should be updated to:
```suggestion
        match, certainty, _ = process.extractOne(query, self.headers.keys())
```

Shom770 commented:
![image](https://user-images.githubusercontent.com/82843611/132365732-111156fb-7bda-4fc5-a669-c91f5b3b2061.png)
@brad90four Would you know as to why this is happening?

brad90four commented:
> 
> 
> ![image](https://user-images.githubusercontent.com/82843611/132365732-111156fb-7bda-4fc5-a669-c91f5b3b2061.png)
> @brad90four Would you know as to why this is happening?

Nope, I went ahead and marked as draft if it is not working. I will need to dive deeper into debugging, I am currently away from my testing setup.

Shom770 commented:
> > ![image](https://user-images.githubusercontent.com/82843611/132365732-111156fb-7bda-4fc5-a669-c91f5b3b2061.png)
> > @brad90four Would you know as to why this is happening?
> 
> Nope, I went ahead and marked as draft if it is not working. I will need to dive deeper into debugging, I am currently away from my testing setup.

I found the issue -- the minimum certainty is too high.

Right now it's at 75, and when you do ``.wtf del``, it finds the string, but the certainty is only 60%

brad90four commented:
Hello all,

Just clarifying something for this command and the implementation of it. Is the goal to send all of the text from the Markdown or just a link to that section? 

The way the command is now, it is only the link. In `#dev-contrib` I had a short [conversation](https://discord.com/channels/267624335836053506/635950537262759947/884815062454374501) with Jason Terror that the intention was to parse the `README.md` file and send the text of the section back to the user. 

With the size of the text that could be sent I just wanted to make sure that this is the intention. 

# Current Status:
 
![image](https://user-images.githubusercontent.com/42116429/132540685-8d5395c3-0be8-47a8-9dab-7883bac91515.png)

# Future Status (mockup):

`.wtf del`
▶ Stubborn del operation

```py
class SomeClass:
    def __del__(self):
        print("Deleted!")
```
Output: 1.
```py
\>>> x = SomeClass()
\>>> y = x
\>>> del x # this should print "Deleted!"
\>>> del y
Deleted!
```
Phew, deleted at last. You might have guessed what saved __del__ from being called in our first attempt to delete x. Let's add more twists to the example.

2.
```py
\>>> x = SomeClass()
\>>> y = x
\>>> del x
\>>> y # check if y exists
<__main__.SomeClass instance at 0x7f98a1a67fc8>
\>>> del y # Like previously, this should print "Deleted!"
\>>> globals() # oh, it didn't. Let's check all our global variables and confirm
Deleted!
{'__builtins__': <module '__builtin__' (built-in)>, 'SomeClass': <class __main__.SomeClass at 0x7f98a1a5f668>, '__package__': None, '__name__': '__main__', '__doc__': None}
```
Okay, now it's deleted 😕
💡 Explanation:

    del x doesn’t directly call x.__del__().
    When del x is encountered, Python deletes the name x from current scope and decrements by 1 the reference count of the object x referenced. __del__() is called only when the object's reference count reaches zero.
    In the second output snippet, __del__() was not called because the previous statement (>>> y) in the interactive interpreter created another reference to the same object (specifically, the _ magic variable which references the result value of the last non None expression on the REPL), thus preventing the reference count from reaching zero when del y was encountered.
    Calling globals (or really, executing anything that will have a non None result) caused _ to reference the new result, dropping the existing reference. Now the reference count reached 0 and we can see "Deleted!" being printed (finally!).

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,159 @@
+    @tasks.loop(hours=1)
+    async def get_wtf_python_readme(self) -> None:
+        """Gets the content of README.md from the WTF Python Repository."""
+        failed_tries = 0
+
+        for x in range(FETCH_TRIES):
+            async with self.bot.http_session.get(f"{WTF_PYTHON_RAW_URL}README.md") as resp:
+                log.trace("Fetching the latest WTF Python README.md")
+
+                if resp.status == 200:
+                    raw = await resp.text()
+                    self.parse_readme(raw)
+                    log.debug(
+                        "Successfully fetched the latest WTF Python README.md, "
+                        "breaking out of retry loop"
+                    )
+                    break
+
+                else:
+                    failed_tries += 1
+                    log.debug(
+                        "Failed to get latest WTF Python README.md on try "
+                        f"{x}/{FETCH_TRIES}. Status code {resp.status}"
+                    )
+
+        if failed_tries == 3:
+            log.error("Couldn't fetch WTF Python README.md after 3 tries, unloading extension.")
+            action = Action.UNLOAD
+        else:
+            action = Action.LOAD
+
+        verb = action.name.lower()
+        ext = "bot.exts.utilities.wtf_python"
+
+        try:
+            action.value(self.bot, ext)
+        except (commands.ExtensionAlreadyLoaded, commands.ExtensionNotLoaded):
+            log.debug(f"Extension `{ext}` is already {verb}ed.")
+        else:
+            log.debug(f"Extension {verb}ed: `{ext}`.")
```
Instead of this, can't we just have the bot save a time of when it was last fetched, then when you use the command and that is older than an hour it will attempt to fetch a new copy?

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,159 @@
+        table_of_contents = re.findall(
+            r"\[👀 Examples\]\(#-examples\)\n([\w\W]*)<!-- tocstop -->", data
+        )[0].split("\n")
+        table_of_contents = list(map(str.strip, table_of_contents))
```
In general it's better to do your best to stay away from overriding variables unless that is it's purpose. Even then you only want to do so when they represent the same thing.

In this case, we can just move the variable into the for loop itself. The benefit of that as well is `map()` being an iterator so we skip exhausting the list.

brad90four replied:
To check that I'm following you here:
From this:
```py
        table_of_contents = re.findall(
            r"\[👀 Examples\]\(#-examples\)\n([\w\W]*)<!-- tocstop -->", data
        )[0].split("\n")
        table_of_contents = list(map(str.strip, table_of_contents))
        log.debug(f"{table_of_contents = }")

        for header in table_of_contents:
```

to this: 
```py
        table_of_contents = re.findall(
            r"\[👀 Examples\]\(#-examples\)\n([\w\W]*)<!-- tocstop -->", data
        )[0].split("\n")
        log.debug(f"{table_of_contents = }")

        for header in list(map(str.strip, table_of_contents)):
```

Bluenix2 replied:
Yes

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,159 @@
+        link to the heading.
+        """
+        table_of_contents = re.findall(
+            r"\[👀 Examples\]\(#-examples\)\n([\w\W]*)<!-- tocstop -->", data
```
It is not awfully apparent to me what these regexes match, are you able to add a comment explaining what they match so that I can see at a glance?

brad90four replied:
Yeah the README is pretty wild with these in the text: https://github.com/satwikkansal/wtfpython/blob/master/README.md#-examples

Bluenix2 replied:
Right so it matches a markdown header and then what's `<!-- tocstop -->`?

brad90four replied:
To be honest, I am not sure. I inherited this PR btw.

But in the raw .md file, the table of contents has that string at the end of the table of contents:

```
# Table of Contents

<!-- Generated using "markdown-toc -i README.md --maxdepth 3"-->

<!-- toc -->

- [Structure of the Examples](#structure-of-the-examples)
    + [▶ Some fancy Title](#-some-fancy-title)
- [Usage](#usage)
- [👀 Examples](#-examples)
  * [Section: Strain your brain!](#section-strain-your-brain)
    + [▶ First things first! *](#-first-things-first-)
    + [▶ Strings can be tricky sometimes](#-strings-can-be-tricky-sometimes)
    + [▶ Be careful with chained operations](#-be-careful-with-chained-operations)
...
...
...
    + [▶ Slowing down `dict` lookups *](#-slowing-down-dict-lookups-)
    + [▶ Bloating instance `dict`s *](#-bloating-instance-dicts-)
    + [▶ Minor Ones *](#-minor-ones-)
- [Contributing](#contributing)
- [Acknowledgements](#acknowledgements)
- [🎓 License](#-license)
  * [Surprise your friends as well!](#surprise-your-friends-as-well)
  * [More content like this?](#more-content-like-this)

<!-- tocstop -->
```

Bluenix2 replied:
Oh I get it now, `<!-- XYZ -->` is of course a comment (HTML-style). Then it stands for "Table Of Contents Stop".

What about a comment like this?
```suggestion
            # Match the start of examples, until the end of the table of contents (toc)
            r"\[👀 Examples\]\(#-examples\)\n([\w\W]*)<!-- tocstop -->", data
```

brad90four replied:
Added that above the `re.findall` line. I'll need to test before I change that to `re.search`

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,159 @@
+        for header in table_of_contents:
+            match = re.findall(r"\[▶ (.*)\]\((.*)\)", header)
+            if match:
+                self.headers.update(
+                    {
+                        match[0][0]: f"{BASE_URL}{match[0][1]}"
+                    }
+                )
```
```suggestion
                self.headers[match[0][0]] = f"{BASE_URL}{match[0][1]}"})
```

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,159 @@
+        table_of_contents = list(map(str.strip, table_of_contents))
+
+        for header in table_of_contents:
+            match = re.findall(r"\[▶ (.*)\]\((.*)\)", header)
```
Can't you use [re.search](https://docs.python.org/3/library/re.html#re.search) instead? Considering you only use the first match.

brad90four replied:
The way the program is currently written, the `parse_readme` function creates a dictionary of the Table of Content's headers, as well as the hyperlink to them. This gets called wtih `fetch_readme`, which is called in the `__init__`.

Once this dictionary is created, the `fuzzy_match_header` function uses the `rapidfuzz.process.extractOne` method with the user search term and this created dictionary.

If I change that to re.search, then I probably wouldn't be leveraging the fuzzy matching properly with `rapidfuzz`. Is it worth the re-write for that?

Bluenix2 replied:
I don't think I understand, I simply suggested changing this `re.findall` to `re.search` because you only grab the first result from the "findall".

brad90four replied:
Yeah, I tested it out and it still worked. Fixed it!

brad90four replied:
Whoa I definitely thought that was updated.

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,159 @@
+                description=f"[Go to Repository Section]({self.headers[match]})",
+            )
+            embed.set_thumbnail(url=f"{WTF_PYTHON_RAW_URL}images/logo.png")
+        await ctx.send(embed=embed)
```
```suggestion

        await ctx.send(embed=embed)
```

brad90four commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,159 @@
+        self.bot = bot
+        self.get_wtf_python_readme.start()
+
+        self.headers: Dict[str] = dict()
```
It should be the header of the README.md that matches the user's query, followed by the link to that header. 
This is the README that is being searched against:
https://github.com/satwikkansal/wtfpython/blob/master/README.md

An example header: 
### ▶ Evaluation time discrepancy

And the generated link: 
https://github.com/satwikkansal/wtfpython/blob/master/README.md#-evaluation-time-discrepancy

Bluenix2 replied:
So `str`?

brad90four replied:
Yessir

Bluenix2 replied:
Additionally, after #802 the repository now fully uses `dict` over `typing.Dict`.
```suggestion
        self.headers: dict[str, str] = dict()
```

D0rs4n commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,170 @@
+        header_stop = self.raw.find("\n---\n", header_start)
+        content = self.raw[header_start:header_stop]
+        file_name = "WTF_content.md"
+        with open(file_name, "w") as f:
```
I'm not an expert of async programming, but I believe this might be a blocking call. Normally, I would consider using something like aiofiles, however using something like `io.StringIO` might fit your needs better. You can use it with `discord.File` since it's a file-like object.

brad90four replied:
Resolved with 5dcc43f and 917de69

brad90four commented:
Hey all, 
I think I will just stick with sending the link to the relevant section in the WTF readme file. I don't see a way forward with the file preview of the .md or .txt file.
If that is ok, I will remove the `readme_file` command and the `embed_file` variable.

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,154 @@
+    @commands.command(aliases=("wtf", "WTF"))
+    async def wtf_python(self, ctx: commands.Context, *, query: str) -> None:
+        """
+        Search WTF python.
```
```suggestion
        Search WTF Python repository.
```

brad90four commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,154 @@
+    @tasks.loop(hours=1)
+    async def fetch_readme(self) -> None:
+        """Gets the content of README.md from the WTF Python Repository."""
+        failed_tries = 0
+
+        for x in range(FETCH_TRIES):
+            async with self.bot.http_session.get(f"{WTF_PYTHON_RAW_URL}README.md") as resp:
+                log.trace("Fetching the latest WTF Python README.md")
+
+                if resp.status == 200:
+                    raw = await resp.text()
+                    self.parse_readme(raw)
+                    log.debug(
+                        "Successfully fetched the latest WTF Python README.md, "
+                        "breaking out of retry loop"
+                    )
+                    break
+
+                else:
+                    failed_tries += 1
+                    log.debug(
+                        "Failed to get latest WTF Python README.md on try "
+                        f"{x}/{FETCH_TRIES}. Status code {resp.status}"
+                    )
+
+        if failed_tries == 3:
+            log.error("Couldn't fetch WTF Python README.md after 3 tries, unloading extension.")
+            action = Action.UNLOAD
+        else:
+            action = Action.LOAD
+
+        verb = action.name.lower()
+        ext = "bot.exts.utilities.wtf_python"
+
+        try:
+            action.value(self.bot, ext)
+        except (commands.ExtensionAlreadyLoaded, commands.ExtensionNotLoaded):
+            log.debug(f"Extension `{ext}` is already {verb}ed.")
+        else:
+            log.debug(f"Extension {verb}ed: `{ext}`.")
```
Yeah I'm getting stuck on this one. I haven't used `asyncio` yet so maybe that is what is tripping me up. 
Also, how can I get the "last" time the cog was loaded instead of just calling it immediately?

Here is what I have so far:

```py
class WTFPython(commands.Cog):
    """Cog that allows getting WTF Python entries from the WTF Python repository."""

    def __init__(self, bot: Bot):
        self.bot = bot
        self.headers: dict[str, str] = dict()
        self.last_fetched = datetime.datetime.now()  # this is just the current time, how to access last load?
        log.debug(f"{self.last_fetched = }")

        if self.last_fetched + datetime.timedelta(seconds=10) < datetime.datetime.now():  # this will never be true currently
            log.debug("Last loading time was within x amount of time.")
            asyncio.create_task(self.fetch_readme)  # is this correct?

    # cut out most everything from the function, this is what is left
    async def fetch_readme(self) -> None:
        """Gets the content of README.md from the WTF Python Repository."""
        async with self.bot.http_session.get(f"{WTF_PYTHON_RAW_URL}README.md") as resp:
            log.trace("Fetching the latest WTF Python README.md")
            if resp.status == 200:
                raw = await resp.text()
                self.parse_readme(raw)
```

Bluenix2 replied:
I am thinking something more like this:
```python

README_REFRESH = 60  # Amount of minutes

class WTFPython(commands.Cog):
    """Cog that allows getting WTF Python entries from the WTF Python repository."""

    def __init__(self, bot: Bot):
        self.bot = bot
        self.headers: dict[str, str] = dict()
        self.last_fetched = datetime.datetime.now()
        log.debug(f"{self.last_fetched = }")

        asyncio.create_task(self.fetch_readme)  # Start a first task

    async def fetch_readme(self) -> None:
        """Gets the content of README.md from the WTF Python Repository."""
        refresh = self.last_fetched + datetime.timedelta(minutes=README_REFRESH)

        if refresh > datetime.datetime.now() and self.headers:  # We also want to fetch if we haven't previously
            return  # Our cache should be up-to-date

        async with self.bot.http_session.get(f"{WTF_PYTHON_RAW_URL}README.md") as resp:
            log.trace("Fetching the latest WTF Python README.md")
            if resp.status == 200:
                raw = await resp.text()
                self.parse_readme(raw)
```
You then call `asyncio.create_task(self.fetch_readme)` every time someone runs the command.

brad90four replied:
Whoa that is awesome! Thanks for the super clear example.

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,121 @@
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.headers: dict[str, str] = dict()
```
```suggestion
        self.headers: dict[str, str] = {}
```

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,121 @@
+        self.bot = bot
+        self.headers: dict[str, str] = dict()
+        self.last_fetched = datetime.datetime.now()
+        log.debug(f"{self.last_fetched = }")
```
```suggestion
```
Please make sure to remove the debug calls that shouldn't make it into main.

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,121 @@
+        with 100 being a perfect match.
+        """
+        match, certainty, _ = rapidfuzz.process.extractOne(query, self.headers.keys())
+        log.debug(f"{match = }, {certainty = }")
```
```suggestion
```

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,122 @@
+                colour=constants.Colours.dark_green,
+                description=f"[Go to Repository Section]({self.headers[match]})",
+            )
+            embed.set_thumbnail(url=f"{WTF_PYTHON_RAW_URL}images/logo.png")
```
This looks poor on dark mode:

![image](https://user-images.githubusercontent.com/38372706/135146203-33ea193d-0f73-42cd-b30a-64109de34bdc.png)

brad90four replied:
Since the image is a .png, how can I change the embed background but only for dark mode?

Bluenix2 replied:
You'll have to find a picture that works better on both platforms, there's no way to differentiate it like that.

brad90four replied:
I could save it as a .jpg and add it to the resources folder?

Bluenix2 replied:
Because of the license ([see here](https://github.com/satwikkansal/wtfpython/blob/master/LICENSE)) we actually could do that afaik.

For legal reasons I am not impliying anything and this is not legal advice ;)

brad90four replied:
How about no image at all?

brad90four replied:
Doesn't look to bad to me:

![image](https://user-images.githubusercontent.com/42116429/135197578-a981403c-a1ee-45a1-a846-3dc654207cda.png)

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,122 @@
+            embed = Embed(
+                title=f"WTF Python Search Result For {query}",
+                colour=constants.Colours.dark_green,
+                description=f"[Go to Repository Section]({self.headers[match]})",
```
The hyperlink doesn't work for me:
![image](https://user-images.githubusercontent.com/38372706/135146347-672247a9-8d32-4eae-8bf4-d0f15fd1e631.png)

Additionally, the URL leads to a 404: https://github.com/satwikkansal/wtfpythonWild

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,122 @@
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.headers: dict[str, str] = {}
+        self.last_fetched = datetime.datetime.now()
```
This is no longer used since you decided to use discord.py's tasks.

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,122 @@
+    @tasks.loop(minutes=README_REFRESH)
+    async def fetch_readme(self) -> None:
+        """Gets the content of README.md from the WTF Python Repository."""
+        refresh = self.last_fetched + datetime.timedelta(minutes=README_REFRESH)
+
+        if refresh > datetime.datetime.now() and self.headers:
+            return  # Cache should be up-to-date
```
Dead code

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,122 @@
+"""
+
+MINIMUM_CERTAINTY = 50
+README_REFRESH = 60  # minutes between fetch calls
```
Unneccessary, discord.py now takes care of this

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,122 @@
+        match = self.fuzzy_match_header(query)
+        if match:
+            embed = Embed(
+                title=f"WTF Python Search Result For {query}",
```
```suggestion
                title=f"WTF Python search result for '{query}'",
```
Additionally, isn't this better added in the description and the title can be something more general?

brad90four replied:
How does this look? 

![image](https://user-images.githubusercontent.com/42116429/135174829-9ed5c78f-ac42-46d2-a5fb-60e9f2384f94.png)

Bluenix2 replied:
I like that, but the capitalization is a bit over-the-top (just do `Search results for 'del'`)

Bluenix2 commented on bot/exts/utilities/wtf_python.py:
```diff
@@ -0,0 +1,118 @@
+        match = self.fuzzy_match_header(query)
+        if match:
+            log.debug(f"{self.headers[match]}")
+            embed = Embed(
+                title="WTF Python?!",
+                colour=constants.Colours.dark_green,
+                description=f"""Search result for '{query}'
+                [Go to Repository Section]({self.headers[match]})""",
+            )
```
Can this embed clarify that an option was found and what it is?

brad90four replied:
More than the quoted user query? Maybe the `self.headers[match]` to show exactly what the link is taking the user to?

Bluenix2 replied:
Yeah something like that. So that it shows that it found for example "wtf imports" or similar.

brad90four replied:
This is what it looks like after 2197fc6:
![image](https://user-images.githubusercontent.com/42116429/135936483-90fdd043-f469-4691-ac30-be4ec34f9adb.png)

Bluenix2 replied:
I like that, thank you Brad.

Bluenix2 reviewed (commented):
Can we have the command have some minimum threshold?

I tried it with the start of the Bee Movie script, first few seconds of Nicki Minaj's superbass, first page of the Bible, the first sentence of the FitnessGram PACER Test, a paragraph about the MRSA bacteria, the MIT license and your name....

but it wasn't until I brought out the morse code of "Hello World" and Never Gonna Give You Up music video link that I got a failure.
